sdks/python/apache_beam/coders/coder_impl.py

```diff
@@ -500,7 +500,9 @@ def encode_special_deterministic(self, value, stream):
         self.encode_to_stream(value.value, stream, True)
       except Exception as e:
         raise TypeError(self._deterministic_encoding_error_msg(value)) from e
-    elif hasattr(value, "__getstate__"):
+    elif (hasattr(value, "__getstate__") and
+          # https://github.com/apache/beam/issues/33020
+          type(value).__reduce__ == object.__reduce):
       if not hasattr(value, "__setstate__"):
```

Comment:
are we sure this works?
```
class MyClass:
    def __init__(self, value):
        self.value = value

    def __reduce__(self):
        return (self.__class__, (self.value,))

tt = MyClass(1)
print(object.__reduce__ == type(tt).__reduce__)
```
This also returns False. I might misunderstand how this works.

Reply:
We want this to return `False` as MyClass has a custom `__reduce__` method (meaning relying on `__set/getstate__` is not safe).

Reply:
Well, thanks goes to you for hunting down the root cause; it was not obvious at all what was going on here :). I'll merge once tests pass.